File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int helper(TreeNode root,int[] dia){
        if(root==null) return 0;
        int l = helper(root.left,dia);
        int r = helper(root.right,dia);
        if(l+r>dia[0]) dia[0] = l+r;
        return Math.max(l,r)+1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null) return 0;
        // int dia = 0;
        int[] dia = new int[1];
        helper(root,dia);
        return dia[0];
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
    int helper(TreeNode* root,int& ans){
        if(!root) return 0;
        int left = helper(root->left,ans);
        int right = helper(root->right,ans);
        ans = max(ans,left+right);
        return max(left,right)+1;
    }
    
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        int ans = 0;
        helper(root,ans);
        return ans;
    }
};